docs/source/advanced/advanced_gpu.rst

```diff
@@ -0,0 +1,571 @@
+Sharded Training allows you to maintain GPU scaling efficiency, whilst reducing memory overhead drastically. In short, expect near-normal linear scaling (if your network allows), and significantly reduced memory usage when training large models.
+
+Sharded Training still utilizes Data Parallel Training under the hood, except optimizer states and gradients are sharded across GPUs.
+This means the memory overhead per GPU is lower, as each GPU only has to maintain a partition of your optimizer state and gradients.
+
+The benefits vary by model and parameter sizes, but we've recorded up to a 63% memory reduction per GPU allowing us to double our model sizes. Because of efficient communication,
+these benefits in multi-GPU setups are almost free and throughput scales well with multi-node setups.
+
+It is highly recommended to use Sharded Training in multi-GPU environments where memory is limited, or where training larger models are beneficial (500M+ parameter models).
+A technical note: as batch size scales, storing activations for the backwards pass becomes the bottleneck in training. As a result, sharding optimizer state and gradients becomes less impactful.
+Use :ref:`fairscale-activation-checkpointing` to see even more benefit at the cost of some throughput.
+
+To use Sharded Training, you need to first install FairScale using the command below.
+
+.. code-block:: bash
+
+    pip install fairscale
+
+
+.. code-block:: python
+
+    # train using Sharded DDP
+    trainer = Trainer(plugins='ddp_sharded')
+
+Sharded Training can work across all DDP variants by adding the additional ``--plugins ddp_sharded`` flag.
+
+Internally we re-initialize your optimizers and shard them across your machines and processes. We handle all communication using PyTorch distributed, so no code changes are required.
+
+----------
+
+.. _fairscale-activation-checkpointing:
+
+FairScale Activation Checkpointing
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Activation checkpointing frees activations from memory as soon as they are not needed during the forward pass. They are then re-computed for the backwards pass as needed.
+
+FairScales' checkpointing wrapper also handles batch norm layers correctly unlike the PyTorch implementation, ensuring stats are tracked correctly due to the multiple forward passes.
+
+This saves memory when training larger models however requires wrapping modules you'd like to use activation checkpointing on. See `here <https://fairscale.readthedocs.io/en/latest/api/nn/misc/checkpoint_activations.html>`__ for more information.
+
+.. code-block:: python
+
+    from pytorch_lightning import Trainer
+    from fairscale.nn import checkpoint_wrapper
+
+
+    class MyModel(pl.LightningModule):
+        def __init__(self):
+            # Wrap layer using checkpoint_wrapper
+            linear_layer = checkpoint_wrapper(nn.Linear(32, 32))
+            self.block = nn.Sequential(linear_layer, nn.ReLU())
```

Comment:
would it make sense to either checkpoint the ReLU as well or checkpoint the whole block? The ReLU should be even cheaper to compute and free the same amount of memory.